packages/next-server/lib/router/router.js

```diff
@@ -154,39 +154,42 @@ export default class Router {
 
     this.abortComponentLoad(as)
 
+    const { pathname: asPathname, query: asQuery } = parse(as, true)
+    const { pathname, query } = parse(url, true)
+    const route = toRoute(pathname)
+
+    // Reuse existing routeInfo when possible.
+    let routeInfo = this.components[route] || null
+
     // If the url change is only related to a hash change
     // We should not proceed. We should only change the state.
     if (this.onlyAHashChange(as)) {
       Router.events.emit('hashChangeStart', as)
       this.changeState(method, url, as)
       this.scrollToHash(as)
+
+      // record hash change
+      const hash = window.location.hash.substring(1)
+      this.set(route, pathname, query, as, { ...routeInfo, hash })
```

Comment:
This call here is the fix. Previously, when `onlyAHashChange` was called, `this.asPath` was undefined because it was never set inside this conditional.